src/compiler/scala/tools/nsc/Global.scala

```diff
@@ -399,31 +408,86 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
     }
 
-    final def withCurrentUnit(unit: CompilationUnit)(task: => Unit): Unit = {
-      if ((unit ne null) && unit.exists)
-        lastSeenSourceFile = unit.source
+    // Cleanup method needed for some phases, for example typer
+    def afterUnit(unit: CompilationUnit): Unit = {}
 
-      if (settings.debug && (settings.verbose || currentRun.size < 5))
-        inform("[running phase " + name + " on " + unit + "]")
-      if (!cancelled(unit)) {
-        currentRun.informUnitStarting(this, unit)
-        try withCurrentUnitNoLog(unit)(task)
-        finally currentRun.advanceUnit()
+    def run(): Unit = {
+      assertOnMain()
+
+      if (isDebugPrintEnabled) inform("[running phase " + name + " on " + currentRun.size +  " compilation units]")
+
+      implicit val ec: ExecutionContextExecutor = createExecutionContext()
+
+      /* Every unit is now run in separate `Future`. If given phase is not ran as parallel one
+       * (which is indicated by `isParallel`) it's swill run on the main thread. This is accomplished by
+       * properly modified `ExecutionContext` returned by `createExecutionContext`.
+       */
+      val futures = currentRun.units.collect {
+        case unit if !cancelled(unit) =>
+          Future {
+            asWorkerThread {
+              processUnit(unit)
+              afterUnit(unit)
+              reporter
+            }
+          }
+      }
+
+      /* Dumping messages from unit's `BufferedReporter` to main reporter.
+       * Since we are awaiting for previous units this allows us to retain messages order.
+       */
+      futures.foreach { future =>
+        val workerReporter = Await.result(future, Duration.Inf)
+        if (isParallel) workerReporter.asInstanceOf[BufferedReporter].flushTo(reporter)
       }
     }
 
-    final def withCurrentUnitNoLog(unit: CompilationUnit)(task: => Unit): Unit = {
+    // Used in methods like `compileLate`
+    final def applyPhase(unit: CompilationUnit): Unit = {
+      assertOnWorker()
+      if (!cancelled(unit)) processUnit(unit)
+    }
+
+    private def processUnit(unit: CompilationUnit): Unit = {
+      assertOnWorker()
+
+      /* In worker threads if we are processing units in parallel we want to use temporary `BufferedReporter` for every unit.
+       * Then later we can then keep it until all previous units are processed and then dump all messages to main reporter.
+       */
+      if (isParallel) reporter = new BufferedReporter
+
+      if (isDebugPrintEnabled) inform("[running phase " + name + " on " + unit + "]")
+
       val unit0 = currentUnit
+
       try {
+        if ((unit ne null) && unit.exists) lastSeenSourceFile = unit.source
         currentRun.currentUnit = unit
-        task
+        _synchronizeNames = isParallel
+        apply(unit)
       } finally {
-        //assert(currentUnit == unit)
         currentRun.currentUnit = unit0
+        currentRun.advanceUnit()
+        _synchronizeNames = false
```

Comment:
races!!!!

Reply:
Ah... you are right!
First thing i need to do at Monday is remove all excludes from my tool and replace them with some intelligent check for synchronization etc.